🐛 Destination BigQuery denormalized (MongoDB source): JSON table encountered too many errors, giving up.

## Environment
- **Airbyte version**:  0.33.5-alpha
- **OS Version / Instance**: Container-Optimized OS from Google 
- **Deployment**: Kubernetes
- **Source Connector and version**: mongo v2 0.1.8
- **Destination Connector and version**: BigQuery Denormalized 0.1.10
- **Severity**: High 
- **Step where error happened**: Sync job

## Current Behavior
Sync fails with the error:
```
[35mdestination[0m - 2021-12-07 14:23:12 INFO DefaultAirbyteStreamFactory(lambda$create$0):61 - 2021-12-07 14:23:12 [32mINFO[m i.a.i.d.b.BigQueryRecordConsumer(closeNormalBigqueryStreams):312 - {} - Removing tmp tables...
[35mdestination[0m - 2021-12-07 14:23:12 INFO DefaultAirbyteStreamFactory(lambda$create$0):61 - 2021-12-07 14:23:12 [32mINFO[m i.a.i.d.b.BigQueryRecordConsumer(closeNormalBigqueryStreams):314 - {} - Finishing destination process...completed
[35mdestination[0m - 2021-12-07 14:23:12 ERROR LineGobbler(voidCall):82 - Exception in thread "main" java.lang.RuntimeException: com.google.cloud.bigquery.BigQueryException: Error while reading data, error message: JSON table encountered too many errors, giving up. Rows: 1; errors: 1. Please look into the errors[] collection for more details.
[35mdestination[0m - 2021-12-07 14:23:12 ERROR LineGobbler(voidCall):82 - 	at io.airbyte.integrations.destination.bigquery.BigQueryRecordConsumer.lambda$closeNormalBigqueryStreams$6(BigQueryRecordConsumer.java:289)
[35mdestination[0m - 2021-12-07 14:23:12 ERROR LineGobbler(voidCall):82 - 	at io.airbyte.commons.lang.Exceptions.castCheckedToRuntime(Exceptions.java:52)
[35mdestination[0m - 2021-12-07 14:23:12 ERROR LineGobbler(voidCall):82 - 	at io.airbyte.commons.lang.Exceptions.toRuntime(Exceptions.java:39)
[35mdestination[0m - 2021-12-07 14:23:12 ERROR LineGobbler(voidCall):82 - 	at io.airbyte.integrations.destination.bigquery.BigQueryRecordConsumer.lambda$closeNormalBigqueryStreams$7(BigQueryRecordConsumer.java:279)
[35mdestination[0m - 2021-12-07 14:23:12 ERROR LineGobbler(voidCall):82 - 	at java.base/java.util.HashMap$Values.forEach(HashMap.java:1068)
[35mdestination[0m - 2021-12-07 14:23:12 ERROR LineGobbler(voidCall):82 - 	at io.airbyte.integrations.destination.bigquery.BigQueryRecordConsumer.closeNormalBigqueryStreams(BigQueryRecordConsumer.java:279)
[35mdestination[0m - 2021-12-07 14:23:12 ERROR LineGobbler(voidCall):82 - 	at io.airbyte.integrations.destination.bigquery.BigQueryRecordConsumer.close(BigQueryRecordConsumer.java:169)
[35mdestination[0m - 2021-12-07 14:23:12 ERROR LineGobbler(voidCall):82 - 	at io.airbyte.integrations.destination.bigquery.BigQueryDenormalizedRecordConsumer.close(BigQueryDenormalizedRecordConsumer.java:81)
[35mdestination[0m - 2021-12-07 14:23:12 ERROR LineGobbler(voidCall):82 - 	at io.airbyte.integrations.base.FailureTrackingAirbyteMessageConsumer.close(FailureTrackingAirbyteMessageConsumer.java:62)
[35mdestination[0m - 2021-12-07 14:23:12 ERROR LineGobbler(voidCall):82 - 	at io.airbyte.integrations.base.IntegrationRunner.consumeWriteStream(IntegrationRunner.java:152)
[35mdestination[0m - 2021-12-07 14:23:12 ERROR LineGobbler(voidCall):82 - 	at io.airbyte.integrations.base.IntegrationRunner.run(IntegrationRunner.java:128)
[35mdestination[0m - 2021-12-07 14:23:12 ERROR LineGobbler(voidCall):82 - 	at io.airbyte.integrations.destination.bigquery.BigQueryDenormalizedDestination.main(BigQueryDenormalizedDestination.java:209)
[35mdestination[0m - 2021-12-07 14:23:12 ERROR LineGobbler(voidCall):82 - Caused by: com.google.cloud.bigquery.BigQueryException: Error while reading data, error message: JSON table encountered too many errors, giving up. Rows: 1; errors: 1. Please look into the errors[] collection for more details.
[35mdestination[0m - 2021-12-07 14:23:12 ERROR LineGobbler(voidCall):82 - 	at com.google.cloud.bigquery.Job.reload(Job.java:411)
[35mdestination[0m - 2021-12-07 14:23:12 ERROR LineGobbler(voidCall):82 - 	at com.google.cloud.bigquery.Job.waitFor(Job.java:248)
[35mdestination[0m - 2021-12-07 14:23:12 ERROR LineGobbler(voidCall):82 - 	at io.airbyte.integrations.destination.bigquery.BigQueryRecordConsumer.lambda$closeNormalBigqueryStreams$6(BigQueryRecordConsumer.java:282)
[35mdestination[0m - 2021-12-07 14:23:12 ERROR LineGobbler(voidCall):82 - 	... 11 more
2021-12-07 14:23:12 INFO KubePodProcess(getReturnCode):583 - Exit code for pod destination-bigquery-denormalized-sync-27-0-xpmjg is 0
2021-12-07 14:23:12 INFO KubePodProcess(exitValue):598 - Closed all resources for pod destination-bigquery-denormalized-sync-27-0-xpmjg
2021-12-07 14:23:12 INFO DefaultReplicationWorker(run):141 - Destination thread complete.

```

yet the sync is marked as _Succeeded_ with no data in destination.

## Expected Behavior
1. Sync is not marked as succeeded when it failed.
2. Bug is fixed or the error message would explain what is wrong with the source data.

## Logs

## Steps to Reproduce
1. set-up connectors
2. run sync

## Are you willing to submit a PR?
no

--- a/airbyte-config/init/src/main/resources/seed/source_definitions.yaml
+++ b/airbyte-config/init/src/main/resources/seed/source_definitions.yaml
@@ -481,7 +481,7 @@
 - name: MongoDb
   sourceDefinitionId: b2e713cd-cc36-4c0a-b5bd-b47cb8a0561e
   dockerRepository: airbyte/source-mongodb-v2
-  dockerImageTag: 0.1.13
+  dockerImageTag: 0.1.14
   documentationUrl: https://docs.airbyte.io/integrations/sources/mongodb-v2
   icon: mongodb.svg
   sourceType: database

--- a/airbyte-config/init/src/main/resources/seed/source_specs.yaml
+++ b/airbyte-config/init/src/main/resources/seed/source_specs.yaml
@@ -5155,7 +5155,7 @@
               path_in_connector_config:
               - "credentials"
               - "client_secret"
-- dockerImage: "airbyte/source-mongodb-v2:0.1.13"
+- dockerImage: "airbyte/source-mongodb-v2:0.1.14"
   spec:
     documentationUrl: "https://docs.airbyte.io/integrations/sources/mongodb-v2"
     changelogUrl: "https://docs.airbyte.io/integrations/sources/mongodb-v2"

--- a/airbyte-db/lib/src/main/java/io/airbyte/db/mongodb/MongoUtils.java
+++ b/airbyte-db/lib/src/main/java/io/airbyte/db/mongodb/MongoUtils.java
@@ -214,7 +214,7 @@ public class MongoUtils {
   }
 
   /**
-   * If one field has different types in 2 and more documents, the name is transofmed to
+   * If one field has different types in 2 and more documents, the name is transformed to
    * 'name_aibyte_transform'.
    *
    * @param types list with field types

--- a/airbyte-db/lib/src/main/java/io/airbyte/db/mongodb/MongoUtils.java
+++ b/airbyte-db/lib/src/main/java/io/airbyte/db/mongodb/MongoUtils.java
@@ -20,13 +20,14 @@ import com.mongodb.client.MongoCollection;
 import io.airbyte.commons.json.Jsons;
 import io.airbyte.commons.util.MoreIterators;
 import io.airbyte.db.DataTypeUtils;
+import io.airbyte.protocol.models.CommonField;
 import io.airbyte.protocol.models.JsonSchemaType;
+import io.airbyte.protocol.models.TreeNode;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collections;
-import java.util.HashMap;
+import java.util.HashSet;
 import java.util.List;
-import java.util.Map;
 import org.bson.BsonBinary;
 import org.bson.BsonDateTime;
 import org.bson.BsonDocument;
@@ -91,6 +92,16 @@ public class MongoUtils {
     } catch (final Exception e) {
       LOGGER.error(String.format("Failed to get BsonValue for field type %s", type), e.getMessage());
       return value;
+    }
+  }
+
+  public static CommonField<BsonType> nodeToCommonField(TreeNode<CommonField<BsonType>> node) {
+    CommonField<BsonType> field = node.getData();
+    if (node.hasChildren()) {
+      List<CommonField<BsonType>> subFields = node.getChildren().stream().map(MongoUtils::nodeToCommonField).toList();
+      return new CommonField<>(field.getName(), field.getType(), subFields);
+    } else {
+      return new CommonField<>(field.getName(), field.getType());
     }
   }
 
@@ -188,21 +199,52 @@ public class MongoUtils {
    * @param collection mongo collection
    * @return map of unique fields and its type
    */
-  public static Map<String, BsonType> getUniqueFields(final MongoCollection<Document> collection) {
+  public static List<TreeNode<CommonField<BsonType>>> getUniqueFields(final MongoCollection<Document> collection) {
+    var allkeys = new HashSet<>(getFieldsName(collection));
 
-    Map<String, BsonType> result = new HashMap<>();
-    var allkeys = getFieldsName(collection);
-    allkeys.forEach(key -> {
+    return allkeys.stream().map(key -> {
       var types = getTypes(collection, key);
-      addUniqueType(result, key, types);
+      var type = getUniqueType(types);
+      var fieldNode = new TreeNode<>(new CommonField<>(transformName(types, key), type));
+      if (type.equals(DOCUMENT)) {
+        setSubFields(collection, fieldNode, key);
+      }
+      return fieldNode;
+    }).toList();
+  }
+
+  /**
+   * If one field has different types in 2 and more documents, the name is transofmed to
+   * 'name_aibyte_transform'.
+   *
+   * @param types list with field types
+   * @param name field name
+   * @return name
+   */
+  private static String transformName(List<String> types, String name) {
+    return types.size() != 1 ? name + AIRBYTE_SUFFIX : name;
+  }
+
+  private static void setSubFields(final MongoCollection<Document> collection, TreeNode<CommonField<BsonType>> parentNode, String pathToField) {
+    var nestedKeys = getFieldsName(collection, pathToField);
+    nestedKeys.forEach(key -> {
+      var types = getTypes(collection, pathToField + "." + key);
+      var nestedType = getUniqueType(types);
+      var childNode = parentNode.addChild(new CommonField<>(transformName(types, key), nestedType));
+      if (nestedType.equals(DOCUMENT)) {
+        setSubFields(collection, childNode, pathToField + "." + key);
+      }
     });
-    return result;
   }
 
   private static List<String> getFieldsName(MongoCollection<Document> collection) {
+    return getFieldsName(collection, "$ROOT");
+  }
+
+  private static List<String> getFieldsName(MongoCollection<Document> collection, String fieldName) {
     AggregateIterable<Document> output = collection.aggregate(Arrays.asList(
         new Document("$limit", DISCOVER_LIMIT),
-        new Document("$project", new Document("arrayofkeyvalue", new Document("$objectToArray", "$$ROOT"))),
+        new Document("$project", new Document("arrayofkeyvalue", new Document("$objectToArray", "$" + fieldName))),
         new Document("$unwind", "$arrayofkeyvalue"),
         new Document("$group", new Document("_id", null).append("allkeys", new Document("$addToSet", "$arrayofkeyvalue.k")))));
     if (output.cursor().hasNext()) {
@@ -233,19 +275,18 @@ public class MongoUtils {
     return listOfTypes;
   }
 
-  private static void addUniqueType(Map<String, BsonType> map,
-                                    String fieldName,
-                                    List<String> types) {
+  private static BsonType getUniqueType(List<String> types) {
     if (types.size() != 1) {
-      map.put(fieldName + AIRBYTE_SUFFIX, BsonType.STRING);
+      return BsonType.STRING;
     } else {
       var type = types.get(0);
-      map.put(fieldName, getBsonTypeByTypeAlias(type));
+      return getBsonTypeByTypeAlias(type);
     }
   }
 
   private static BsonType getBsonTypeByTypeAlias(String typeAlias) {
     return switch (typeAlias) {
+      case "object" -> BsonType.DOCUMENT;
       case "double" -> BsonType.DOUBLE;
       case "string" -> BsonType.STRING;
       case "objectId" -> BsonType.OBJECT_ID;

--- a/airbyte-integrations/connectors/source-mongodb-v2/Dockerfile
+++ b/airbyte-integrations/connectors/source-mongodb-v2/Dockerfile
@@ -16,5 +16,5 @@ ENV APPLICATION source-mongodb-v2
 
 COPY --from=build /airbyte /airbyte
 
-LABEL io.airbyte.version=0.1.13
+LABEL io.airbyte.version=0.1.14
 LABEL io.airbyte.name=airbyte/source-mongodb-v2

--- a/airbyte-integrations/connectors/source-mongodb-v2/src/main/java/io.airbyte.integrations.source.mongodb/MongoDbSource.java
+++ b/airbyte-integrations/connectors/source-mongodb-v2/src/main/java/io.airbyte.integrations.source.mongodb/MongoDbSource.java
@@ -116,11 +116,7 @@ public class MongoDbSource extends AbstractDbSource<BsonType, MongoDatabase> {
 
     for (final String collectionName : getAuthorizedCollections(database)) {
       final MongoCollection<Document> collection = database.getCollection(collectionName);
-      final Map<String, BsonType> uniqueFields = MongoUtils.getUniqueFields(collection);
-
-      final List<CommonField<BsonType>> fields = uniqueFields.keySet().stream()
-          .map(field -> new CommonField<>(field, uniqueFields.get(field)))
-          .collect(Collectors.toList());
+      final List<CommonField<BsonType>> fields = MongoUtils.getUniqueFields(collection).stream().map(MongoUtils::nodeToCommonField).toList();
 
       // The field name _id is reserved for use as a primary key;
       final TableInfo<CommonField<BsonType>> tableInfo = TableInfo.<CommonField<BsonType>>builder()

--- a/airbyte-integrations/connectors/source-mongodb-v2/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/MongoDbSourceAbstractAcceptanceTest.java
+++ b/airbyte-integrations/connectors/source-mongodb-v2/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/MongoDbSourceAbstractAcceptanceTest.java
@@ -61,7 +61,8 @@ public abstract class MongoDbSourceAbstractAcceptanceTest extends SourceAcceptan
                 Field.of("test_array", JsonSchemaType.ARRAY),
                 Field.of("empty_test", JsonSchemaType.STRING),
                 Field.of("double_test", JsonSchemaType.NUMBER),
-                Field.of("int_test", JsonSchemaType.NUMBER))
+                Field.of("int_test", JsonSchemaType.NUMBER),
+                Field.of("object_test", JsonSchemaType.OBJECT))
                 .withSupportedSyncModes(Lists.newArrayList(SyncMode.INCREMENTAL))
                 .withDefaultCursorField(List.of("_id")))));
   }

--- a/airbyte-integrations/connectors/source-mongodb-v2/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/MongoDbSourceAtlasAcceptanceTest.java
+++ b/airbyte-integrations/connectors/source-mongodb-v2/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/MongoDbSourceAtlasAcceptanceTest.java
@@ -107,7 +107,6 @@ public class MongoDbSourceAtlasAcceptanceTest extends MongoDbSourceAbstractAccep
     // only one stream is expected; the schema that should be ignored
     // must not be included in the retrieved catalog
     assertEquals(1, streams.size());
-    ObjectMapper mapper = new ObjectMapper();
     final AirbyteStream actualStream = streams.get(0);
     assertEquals(CatalogHelpers.fieldsToJsonSchema(FIELDS), actualStream.getJsonSchema());
   }

--- a/airbyte-integrations/connectors/source-mongodb-v2/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/MongoDbSourceAtlasAcceptanceTest.java
+++ b/airbyte-integrations/connectors/source-mongodb-v2/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/MongoDbSourceAtlasAcceptanceTest.java
@@ -5,13 +5,20 @@
 package io.airbyte.integrations.io.airbyte.integration_tests.sources;
 
 import static io.airbyte.db.mongodb.MongoUtils.MongoInstanceType.ATLAS;
+import static org.junit.jupiter.api.Assertions.assertEquals;
 
 import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
 import com.google.common.collect.ImmutableMap;
 import com.mongodb.client.MongoCollection;
 import io.airbyte.commons.json.Jsons;
 import io.airbyte.db.mongodb.MongoDatabase;
 import io.airbyte.integrations.standardtest.source.TestDestinationEnv;
+import io.airbyte.protocol.models.AirbyteCatalog;
+import io.airbyte.protocol.models.AirbyteStream;
+import io.airbyte.protocol.models.CatalogHelpers;
+import io.airbyte.protocol.models.Field;
+import io.airbyte.protocol.models.JsonSchemaType;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.util.List;
@@ -22,6 +29,25 @@ import org.bson.Document;
 public class MongoDbSourceAtlasAcceptanceTest extends MongoDbSourceAbstractAcceptanceTest {
 
   private static final Path CREDENTIALS_PATH = Path.of("secrets/credentials.json");
+
+  protected static final List<Field> SUB_FIELDS = List.of(
+      Field.of("testObject", JsonSchemaType.OBJECT, List.of(
+          Field.of("name", JsonSchemaType.STRING),
+          Field.of("testField1", JsonSchemaType.STRING),
+          Field.of("testInt", JsonSchemaType.NUMBER),
+          Field.of("thirdLevelDocument", JsonSchemaType.OBJECT, List.of(
+              Field.of("data", JsonSchemaType.STRING),
+              Field.of("intData", JsonSchemaType.NUMBER))))));
+
+  protected static final List<Field> FIELDS = List.of(
+      Field.of("id", JsonSchemaType.STRING),
+      Field.of("_id", JsonSchemaType.STRING),
+      Field.of("name", JsonSchemaType.STRING),
+      Field.of("test_aibyte_transform", JsonSchemaType.STRING),
+      Field.of("test_array", JsonSchemaType.ARRAY),
+      Field.of("int_test", JsonSchemaType.NUMBER),
+      Field.of("double_test", JsonSchemaType.NUMBER),
+      Field.of("object_test", JsonSchemaType.OBJECT, SUB_FIELDS));
 
   @Override
   protected void setupEnvironment(final TestDestinationEnv environment) throws Exception {
@@ -56,12 +82,15 @@ public class MongoDbSourceAtlasAcceptanceTest extends MongoDbSourceAbstractAccep
     database = new MongoDatabase(connectionString, DATABASE_NAME);
 
     final MongoCollection<Document> collection = database.createCollection(COLLECTION_NAME);
+    final var objectDocument = new Document("testObject", new Document("name", "subName").append("testField1", "testField1").append("testInt", 10)
+        .append("thirdLevelDocument", new Document("data", "someData").append("intData", 1)));
     final var doc1 = new Document("id", "0001").append("name", "Test")
         .append("test", 10).append("test_array", new BsonArray(List.of(new BsonString("test"), new BsonString("mongo"))))
-        .append("double_test", 100.12).append("int_test", 100);
-    final var doc2 = new Document("id", "0002").append("name", "Mongo").append("test", "test_value").append("int_test", 201);
+        .append("double_test", 100.12).append("int_test", 100).append("object_test", objectDocument);
+    final var doc2 =
+        new Document("id", "0002").append("name", "Mongo").append("test", "test_value").append("int_test", 201).append("object_test", objectDocument);
     final var doc3 = new Document("id", "0003").append("name", "Source").append("test", null)
-        .append("double_test", 212.11).append("int_test", 302);
+        .append("double_test", 212.11).append("int_test", 302).append("object_test", objectDocument);
 
     collection.insertMany(List.of(doc1, doc2, doc3));
   }
@@ -70,6 +99,17 @@ public class MongoDbSourceAtlasAcceptanceTest extends MongoDbSourceAbstractAccep
   protected void tearDown(final TestDestinationEnv testEnv) throws Exception {
     database.getDatabase().getCollection(COLLECTION_NAME).drop();
     database.close();
+  }
+
+  @Override
+  protected void verifyCatalog(AirbyteCatalog catalog) {
+    final List<AirbyteStream> streams = catalog.getStreams();
+    // only one stream is expected; the schema that should be ignored
+    // must not be included in the retrieved catalog
+    assertEquals(1, streams.size());
+    ObjectMapper mapper = new ObjectMapper();
+    final AirbyteStream actualStream = streams.get(0);
+    assertEquals(CatalogHelpers.fieldsToJsonSchema(FIELDS), actualStream.getJsonSchema());
   }
 
 }

--- a/airbyte-integrations/connectors/source-mongodb-v2/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/MongoDbSourceStandaloneAcceptanceTest.java
+++ b/airbyte-integrations/connectors/source-mongodb-v2/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/MongoDbSourceStandaloneAcceptanceTest.java
@@ -5,13 +5,20 @@
 package io.airbyte.integrations.io.airbyte.integration_tests.sources;
 
 import static io.airbyte.db.mongodb.MongoUtils.MongoInstanceType.STANDALONE;
+import static org.junit.jupiter.api.Assertions.assertEquals;
 
 import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
 import com.google.common.collect.ImmutableMap;
 import com.mongodb.client.MongoCollection;
 import io.airbyte.commons.json.Jsons;
 import io.airbyte.db.mongodb.MongoDatabase;
 import io.airbyte.integrations.standardtest.source.TestDestinationEnv;
+import io.airbyte.protocol.models.AirbyteCatalog;
+import io.airbyte.protocol.models.AirbyteStream;
+import io.airbyte.protocol.models.CatalogHelpers;
+import io.airbyte.protocol.models.Field;
+import io.airbyte.protocol.models.JsonSchemaType;
 import java.util.List;
 import org.bson.BsonArray;
 import org.bson.BsonString;
@@ -22,6 +29,25 @@ import org.testcontainers.utility.DockerImageName;
 public class MongoDbSourceStandaloneAcceptanceTest extends MongoDbSourceAbstractAcceptanceTest {
 
   private MongoDBContainer mongoDBContainer;
+
+  private static final List<Field> SUB_FIELDS = List.of(
+      Field.of("testObject", JsonSchemaType.OBJECT, List.of(
+          Field.of("name", JsonSchemaType.STRING),
+          Field.of("testField1", JsonSchemaType.STRING),
+          Field.of("testInt", JsonSchemaType.NUMBER),
+          Field.of("thirdLevelDocument", JsonSchemaType.OBJECT, List.of(
+              Field.of("data", JsonSchemaType.STRING),
+              Field.of("intData", JsonSchemaType.NUMBER))))));
+
+  private static final List<Field> FIELDS = List.of(
+      Field.of("id", JsonSchemaType.STRING),
+      Field.of("_id", JsonSchemaType.STRING),
+      Field.of("name", JsonSchemaType.STRING),
+      Field.of("test_aibyte_transform", JsonSchemaType.STRING),
+      Field.of("test_array", JsonSchemaType.ARRAY),
+      Field.of("int_test", JsonSchemaType.NUMBER),
+      Field.of("double_test", JsonSchemaType.NUMBER),
+      Field.of("object_test", JsonSchemaType.OBJECT, SUB_FIELDS));
 
   @Override
   protected void setupEnvironment(final TestDestinationEnv environment) throws Exception {
@@ -48,12 +74,15 @@ public class MongoDbSourceStandaloneAcceptanceTest extends MongoDbSourceAbstract
     database = new MongoDatabase(connectionString, DATABASE_NAME);
 
     final MongoCollection<Document> collection = database.createCollection(COLLECTION_NAME);
+    final var objectDocument = new Document("testObject", new Document("name", "subName").append("testField1", "testField1").append("testInt", 10)
+        .append("thirdLevelDocument", new Document("data", "someData").append("intData", 1)));
     final var doc1 = new Document("id", "0001").append("name", "Test")
         .append("test", 10).append("test_array", new BsonArray(List.of(new BsonString("test"), new BsonString("mongo"))))
-        .append("double_test", 100.12).append("int_test", 100);
-    final var doc2 = new Document("id", "0002").append("name", "Mongo").append("test", "test_value").append("int_test", 201);
+        .append("double_test", 100.12).append("int_test", 100).append("object_test", objectDocument);
+    final var doc2 =
+        new Document("id", "0002").append("name", "Mongo").append("test", "test_value").append("int_test", 201).append("object_test", objectDocument);
     final var doc3 = new Document("id", "0003").append("name", "Source").append("test", null)
-        .append("double_test", 212.11).append("int_test", 302);
+        .append("double_test", 212.11).append("int_test", 302).append("object_test", objectDocument);
 
     collection.insertMany(List.of(doc1, doc2, doc3));
   }
@@ -62,6 +91,17 @@ public class MongoDbSourceStandaloneAcceptanceTest extends MongoDbSourceAbstract
   protected void tearDown(final TestDestinationEnv testEnv) throws Exception {
     database.close();
     mongoDBContainer.close();
+  }
+
+  @Override
+  protected void verifyCatalog(AirbyteCatalog catalog) {
+    final List<AirbyteStream> streams = catalog.getStreams();
+    // only one stream is expected; the schema that should be ignored
+    // must not be included in the retrieved catalog
+    assertEquals(1, streams.size());
+    ObjectMapper mapper = new ObjectMapper();
+    final AirbyteStream actualStream = streams.get(0);
+    assertEquals(CatalogHelpers.fieldsToJsonSchema(FIELDS), actualStream.getJsonSchema());
   }
 
 }

--- a/airbyte-integrations/connectors/source-mongodb-v2/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/MongoDbSourceStandaloneAcceptanceTest.java
+++ b/airbyte-integrations/connectors/source-mongodb-v2/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/MongoDbSourceStandaloneAcceptanceTest.java
@@ -99,7 +99,6 @@ public class MongoDbSourceStandaloneAcceptanceTest extends MongoDbSourceAbstract
     // only one stream is expected; the schema that should be ignored
     // must not be included in the retrieved catalog
     assertEquals(1, streams.size());
-    ObjectMapper mapper = new ObjectMapper();
     final AirbyteStream actualStream = streams.get(0);
     assertEquals(CatalogHelpers.fieldsToJsonSchema(FIELDS), actualStream.getJsonSchema());
   }

--- a/airbyte-integrations/connectors/source-relational-db/src/main/java/io/airbyte/integrations/source/relationaldb/AbstractDbSource.java
+++ b/airbyte-integrations/connectors/source-relational-db/src/main/java/io/airbyte/integrations/source/relationaldb/AbstractDbSource.java
@@ -330,7 +330,7 @@ public abstract class AbstractDbSource<DataType, Database extends AbstractDataba
           assertColumnsWithSameNameAreSame(t.getNameSpace(), t.getName(), t.getFields());
           final List<Field> fields = t.getFields()
               .stream()
-              .map(f -> Field.of(f.getName(), getType(f.getType())))
+              .map(this::toField)
               .distinct()
               .collect(Collectors.toList());
           final String fullyQualifiedTableName = getFullyQualifiedTableName(t.getNameSpace(), t.getName());
@@ -341,6 +341,15 @@ public abstract class AbstractDbSource<DataType, Database extends AbstractDataba
               .build();
         })
         .collect(Collectors.toList());
+  }
+
+  protected Field toField(CommonField<DataType> field) {
+    if (getType(field.getType()) == JsonSchemaType.OBJECT && field.getProperties() != null && !field.getProperties().isEmpty()) {
+      var properties = field.getProperties().stream().map(this::toField).toList();
+      return Field.of(field.getName(), getType(field.getType()), properties);
+    } else {
+      return Field.of(field.getName(), getType(field.getType()));
+    }
   }
 
   protected void assertColumnsWithSameNameAreSame(final String nameSpace, final String tableName, final List<CommonField<DataType>> columns) {

--- a/airbyte-protocol/models/src/main/java/io/airbyte/protocol/models/CatalogHelpers.java
+++ b/airbyte-protocol/models/src/main/java/io/airbyte/protocol/models/CatalogHelpers.java
@@ -100,7 +100,13 @@ public class CatalogHelpers {
             .stream()
             .collect(Collectors.toMap(
                 Field::getName,
-                field -> field.getType().getJsonSchemaTypeMap())))
+                field -> {
+                  if (isObjectWithSubFields(field)) {
+                    return fieldsToJsonSchema(field.getSubFields());
+                  } else {
+                    return field.getType().getJsonSchemaTypeMap();
+                  }
+                })))
         .build());
   }
 
@@ -139,6 +145,10 @@ public class CatalogHelpers {
     }
 
     return allFieldNames;
+  }
+
+  private static boolean isObjectWithSubFields(Field field) {
+    return field.getType() == JsonSchemaType.OBJECT && field.getSubFields() != null && !field.getSubFields().isEmpty();
   }
 
 }

--- a/airbyte-protocol/models/src/main/java/io/airbyte/protocol/models/CommonField.java
+++ b/airbyte-protocol/models/src/main/java/io/airbyte/protocol/models/CommonField.java
@@ -4,6 +4,7 @@
 
 package io.airbyte.protocol.models;
 
+import java.util.List;
 import java.util.Objects;
 
 public class CommonField<T> {
@@ -11,9 +12,18 @@ public class CommonField<T> {
   private final String name;
   private final T type;
 
+  private final List<CommonField<T>> properties;
+
   public CommonField(final String name, final T type) {
     this.name = name;
     this.type = type;
+    this.properties = null;
+  }
+
+  public CommonField(final String name, final T type, List<CommonField<T>> properties) {
+    this.name = name;
+    this.type = type;
+    this.properties = properties;
   }
 
   public String getName() {
@@ -35,12 +45,16 @@ public class CommonField<T> {
 
     final CommonField<T> field = (CommonField<T>) o;
     return name.equals(field.name) &&
-        type == field.type;
+        type == field.type && Objects.equals(properties, field.properties);
   }
 
   @Override
   public int hashCode() {
-    return Objects.hash(name, type);
+    return Objects.hash(name, type, properties);
+  }
+
+  public List<CommonField<T>> getProperties() {
+    return properties;
   }
 
 }

--- a/airbyte-protocol/models/src/main/java/io/airbyte/protocol/models/Field.java
+++ b/airbyte-protocol/models/src/main/java/io/airbyte/protocol/models/Field.java
@@ -4,14 +4,31 @@
 
 package io.airbyte.protocol.models;
 
+import java.util.List;
+
 public class Field extends CommonField<JsonSchemaType> {
+
+  private List<Field> subFields;
 
   public Field(final String name, final JsonSchemaType type) {
     super(name, type);
   }
 
+  public Field(final String name, final JsonSchemaType type, List<Field> subFields) {
+    super(name, type);
+    this.subFields = subFields;
+  }
+
   public static Field of(final String name, final JsonSchemaType type) {
     return new Field(name, type);
+  }
+
+  public static Field of(final String name, final JsonSchemaType type, List<Field> properties) {
+    return new Field(name, type, properties);
+  }
+
+  public List<Field> getSubFields() {
+    return subFields;
   }
 
 }

--- a/airbyte-protocol/models/src/main/java/io/airbyte/protocol/models/TreeNode.java
+++ b/airbyte-protocol/models/src/main/java/io/airbyte/protocol/models/TreeNode.java
@@ -1,0 +1,56 @@
+/*
+ * Copyright (c) 2021 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.protocol.models;
+
+import java.util.LinkedList;
+import java.util.List;
+
+public class TreeNode<T> {
+
+  private T data;
+  private TreeNode<T> parent;
+  private List<TreeNode<T>> children;
+
+  public TreeNode(T data) {
+    this.data = data;
+    this.children = new LinkedList<>();
+  }
+
+  public TreeNode<T> addChild(T child) {
+    TreeNode<T> childNode = new TreeNode<T>(child);
+    childNode.parent = this;
+    this.children.add(childNode);
+    return childNode;
+  }
+
+  public boolean hasChildren() {
+    return children != null && !children.isEmpty();
+  }
+
+  public T getData() {
+    return data;
+  }
+
+  public void setData(T data) {
+    this.data = data;
+  }
+
+  public TreeNode<T> getParent() {
+    return parent;
+  }
+
+  public void setParent(TreeNode<T> parent) {
+    this.parent = parent;
+  }
+
+  public List<TreeNode<T>> getChildren() {
+    return children;
+  }
+
+  public void setChildren(List<TreeNode<T>> children) {
+    this.children = children;
+  }
+
+}

--- a/airbyte-protocol/models/src/test/java/io/airbyte/protocol/models/CatalogHelpersTest.java
+++ b/airbyte-protocol/models/src/test/java/io/airbyte/protocol/models/CatalogHelpersTest.java
@@ -12,6 +12,7 @@ import com.google.common.collect.Sets;
 import io.airbyte.commons.json.Jsons;
 import io.airbyte.commons.resources.MoreResources;
 import java.io.IOException;
+import java.util.List;
 import java.util.Set;
 import org.junit.jupiter.api.Test;
 
@@ -19,8 +20,41 @@ class CatalogHelpersTest {
 
   @Test
   void testFieldToJsonSchema() {
-    final String expected = "{ \"type\": \"object\", \"properties\": { \"name\": { \"type\": \"string\" } } } ";
-    final JsonNode actual = CatalogHelpers.fieldsToJsonSchema(Field.of("name", JsonSchemaType.STRING));
+    final String expected = """
+                                {
+                                  "type": "object",
+                                  "properties": {
+                                    "name": {
+                                      "type": "string"
+                                    },
+                                    "test_object": {
+                                      "type": "object",
+                                      "properties": {
+                                        "thirdLevelObject": {
+                                          "type": "object",
+                                          "properties": {
+                                            "data": {
+                                              "type": "string"
+                                            },
+                                            "intData": {
+                                              "type": "number"
+                                            }
+                                          }
+                                        },
+                                        "name": {
+                                          "type": "string"
+                                        }
+                                      }
+                                    }
+                                  }
+                                }
+                            """;
+    final JsonNode actual = CatalogHelpers.fieldsToJsonSchema(Field.of("name", JsonSchemaType.STRING),
+        Field.of("test_object", JsonSchemaType.OBJECT, List.of(
+            Field.of("name", JsonSchemaType.STRING),
+            Field.of("thirdLevelObject", JsonSchemaType.OBJECT, List.of(
+                Field.of("data", JsonSchemaType.STRING),
+                Field.of("intData", JsonSchemaType.NUMBER))))));
 
     assertEquals(Jsons.deserialize(expected), actual);
   }

--- a/docs/integrations/sources/mongodb-v2.md
+++ b/docs/integrations/sources/mongodb-v2.md
@@ -102,6 +102,7 @@ For more information regarding configuration parameters, please see [MongoDb Doc
 
 | Version | Date | Pull Request | Subject |
 | :--- | :--- | :--- | :--- |
+| 0.1.14 | 2022-05-05 | [12428](https://github.com/airbytehq/airbyte/pull/12428) | JsonSchema: Add properties to fields with type 'object' |
 | 0.1.13  | 2022-02-21 | [10276](https://github.com/airbytehq/airbyte/pull/10276) | Create a custom codec registry to handle DBRef MongoDB objects |
 | 0.1.12 | 2022-02-14 | [10256](https://github.com/airbytehq/airbyte/pull/10256) | (unpublished) Add `-XX:+ExitOnOutOfMemoryError` JVM option |
 | 0.1.11 | 2022-01-10 | [9238](https://github.com/airbytehq/airbyte/pull/9238) | Return only those collections for which the user has privileges |

